File: y2017/rust/src/bin/day7.rs
use std::fs;
use std::collections::{HashMap, HashSet};
use regex::Regex;

fn even_weight(
    at: &String,
    supporting: &HashMap<String, (u32, Vec<String>)>
) -> (u32, Option<u32>) {
    let weight = supporting[at].0;
    if supporting[at].1.is_empty() {
        return (weight, None)
    }

    let mut branches: HashMap<u32, Vec<u32>> = HashMap::new();
    let mut branch_res = Vec::new();
    let mut total = weight;
    for s in &supporting[at].1 {
        let (b_weight, b_res) = even_weight(s, supporting);
        branches.entry(b_weight).or_default().push(supporting[s].0);
        branch_res.push(b_res);
        total += b_weight;
    }

    if let Some(changed) = branch_res.iter().find(|r| r.is_some()) {
        assert!(branches.len() <= 1);
        return (total, *changed);
    }
    match branches.len() {
        1 => (total, None),
        2 => {
            let weights: Vec<&u32> = branches.keys().collect();
            let first = branches[weights[0]].len();
            let second = branches[weights[1]].len();
            assert_ne!((first != 1), (second != 1));

            let bad = branches[weights[if first < second { 0 } else { 1 }]][0];
            let mut change = (weights[1] - weights[0]) as i32;
            change *= if first < second { 1 } else { -1 };
            ((total as i32 + change) as u32, Some((bad as i32 + change) as u32))
        },
        _ => panic!("this is bad, very bad")
    }
}

fn main() {
    let tower = fs::read_to_string("../input/day7.txt").expect("bruh");

    let prog_fmt = Regex::new(r"([a-z]+) \((\d+)\) -> ([a-z,\s]+)").unwrap();
    let prog_leaf_fmt = Regex::new(r"([a-z]+) \((\d+)\)").unwrap();

    let mut supporting = HashMap::new();
    for t in tower.lines() {
        if let Some(m) = prog_fmt.captures(t) {
            let s = m[3].split(",")
                .map(|t| t.trim().to_string())
                .collect::<Vec<String>>();
            supporting.insert(
                m[1].to_string(),
                (m[2].parse::<u32>().unwrap(), s)
            );
        } else if let Some(m) = prog_leaf_fmt.captures(t) {
            supporting.insert(m[1].to_string(),
                              (m[2].parse::<u32>().unwrap(), Vec::new())
            );
        }
    }

    let mut not_bottom = HashSet::new();
    for (_, (_, above)) in &supporting {
        not_bottom.extend(above.into_iter());
    }
    for t in supporting.keys() {
        if !not_bottom.contains(t) {
            println!("bottom of tower: {t}");
            let new_weight = even_weight(t, &supporting).1.unwrap();
            println!("adjusted program weight: {}", new_weight);
            break;
        }
    }
}

File: y2017/nim/day7.nim
import strutils
import sequtils
import sugar
import re
import tables
import sets
import options

proc evenWeight(
  root: string, supporting: Table[string, (int, seq[string])]
): (int, Option[int]) =
  let info = supporting[root]
  if info[1].len == 0:
    return (info[0], none(int))
  
  var
    branches: Table[int, seq[int]]
    toAdjust = none(int)
  for t in info[1]:
    let (weight, adj) = evenWeight(t, supporting)
    if not (weight in branches):
      branches[weight] = @[]
    branches[weight].add(supporting[t][0])
    if adj.isSome:
      toAdjust = adj
  
  case branches.len:
    of 1:
      for w, a in branches.pairs:
        return (w * a.len + info[0], toAdjust)

    of 2:
      if toAdjust.isSome:
        raise newException(ValueError, "bad tree idk man")
      
      let
        tmp = branches.pairs.toSeq
        bad = if tmp[0][1].len == 1: 0 else: 1
        good = 1 - bad
        badDelta = tmp[good][0] - tmp[bad][0]
        changeTo = tmp[bad][1][0] + badDelta
        newWeight = tmp[good][0] * (tmp[good][1].len + 1)

      return (newWeight + info[0], some(changeTo))
    else:
        raise newException(ValueError, "bad tree idk man")

let tower = readFile("input/day7.txt").splitLines()

let
  progFmt = re"^([a-z]+) \((\d+)\) -> ([a-z,\s]+)$"
  progLeafFmt = re"^([a-z]+) \((\d+)\)$"
var supporting: Table[string, (int, seq[string])]
for t in tower:
  var matches1: array[3, string]  # alright this part is pretty cursed lowk
  var matches2: array[2, string]
  if t.find(progFmt, matches1) >= 0:
    supporting[matches1[0]] = (
      parseInt(matches1[1]),
      matches1[2].split(",").map(s => s.strip())
    )
  elif t.find(progLeafFmt, matches2) >= 0:
    supporting[matches2[0]] = (parseInt(matches2[1]), @[])

var notBottom: HashSet[string]
for _, above in supporting.values:
  above.apply(proc(t: string) = notBottom.incl(t))

for t in supporting.keys:
  if not (t in notBottom):
    echo "bottom of tower: $1" % [t]
    let (_, newWeight) = evenWeight(t, supporting)
    if isSome(newWeight):
      echo "adjusted weight to fix crap: $1" % [$newWeight.get]
    else:
      echo "lmao your input must be wrong the tower's alr balanced"
